wgpu-core/src/device/mod.rs

```diff
@@ -2754,29 +2759,24 @@ impl<A: HalApi> Device<A> {
         };
 
         if validated_stages.contains(wgt::ShaderStages::FRAGMENT) {
-            for (i, state) in color_targets.iter().enumerate() {
-                match io.get(&(i as wgt::ShaderLocation)) {
-                    Some(output) => {
-                        validation::check_texture_format(state.format, &output.ty).map_err(
-                            |pipeline| {
-                                pipeline::CreateRenderPipelineError::ColorState(
-                                    i as u8,
-                                    pipeline::ColorStateError::IncompatibleFormat {
-                                        pipeline,
-                                        shader: output.ty,
-                                    },
-                                )
+            for (i, output) in io.iter() {
+                if let Some(state) = color_targets.get(*i as usize) {
+                    let format = if let Some(st) = state.as_ref() {
+                        st.format
+                    } else {
+                        return Err(
+                            pipeline::CreateRenderPipelineError::InvalidFragmentOutputLocation(*i),
+                        );
+                    };
+                    validation::check_texture_format(format, &output.ty).map_err(|pipeline| {
+                        pipeline::CreateRenderPipelineError::ColorState(
+                            *i as u8,
+                            pipeline::ColorStateError::IncompatibleFormat {
+                                pipeline,
+                                shader: output.ty,
                             },
-                        )?;
-                    }
-                    None if state.write_mask.is_empty() => {}
-                    None => {
-                        log::warn!("Missing fragment output[{}], expected {:?}", i, state,);
-                        return Err(pipeline::CreateRenderPipelineError::ColorState(
-                            i as u8,
-                            pipeline::ColorStateError::Missing,
-                        ));
-                    }
+                        )
+                    })?;
```

Comment:
Suggestion: it might be a bit nicer to just flatten this into a `match` statement:

```
                match color_targets.get(*i as usize) {
                    Some(&Some(ref state)) => {
                        let format = state.format;
                        validation::check_texture_format(format, &output.ty).map_err(|pipeline| {
                            pipeline::CreateRenderPipelineError::ColorState(
                                *i as u8,
                                pipeline::ColorStateError::IncompatibleFormat {
                                    pipeline,
                                    shader: output.ty,
                                },
                            )
                        })?;
                    }
                    Some(&None) => {
                        return Err(
                            pipeline::CreateRenderPipelineError::InvalidFragmentOutputLocation(*i),
                        );
                    }
                    None => {
                        /* possibly missing error check here */
                    },
                }
```